File: src/main/java/chaos/unity/nenggao/Span.java
package chaos.unity.nenggao;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class Span extends AbstractSpan {
    public Span(@NotNull AbstractPosition startPosition, @NotNull AbstractPosition endPosition) {
        super(startPosition, endPosition);
    }

    @Override
    public @NotNull AbstractSpan expand(@Nullable AbstractSpan endSpan) {
        AbstractSpan copied = copy();

        if (endSpan == null)
            return copied;

        if (endSpan.endPosition.line < startPosition.line)
            return copied;
        else if (endSpan.endPosition.line == startPosition.line) {
            if (endSpan.endPosition.pos < startPosition.pos)
                return copied;
        }

        AbstractPosition startPosition = this.startPosition;
        AbstractPosition endPosition = endSpan.endPosition;

        return new Span(startPosition, endPosition);
    }

    @Override
    public @NotNull AbstractSpan copy() {
        return new Span(startPosition, endPosition);
    }
}

File: kotlinImpl/src/main/kotlin/chaos/unity/nengao/kotlin/Span.kt
package chaos.unity.nengao.kotlin

import chaos.unity.nenggao.AbstractPosition
import chaos.unity.nenggao.AbstractSpan
import chaos.unity.nenggao.Span

data class Span(val startPosition: AbstractPosition, val endPosition: AbstractPosition) :
    AbstractSpan(startPosition, endPosition) {
    override fun expand(endSpan: AbstractSpan?): AbstractSpan =
        endSpan?.let {
            if (endSpan.endPosition.line < startPosition.line) return copy()
            else if (endSpan.endPosition.line == startPosition.line && endSpan.endPosition.pos < startPosition.pos) return copy()

            return Span(startPosition, endSpan.endPosition)
        } ?: copy()

    override fun copy(): AbstractSpan =
        copy(startPosition = startPosition, endPosition = endPosition)
}
